drivers/modem/ublox-sara-n310.c

```diff
@@ -0,0 +1,526 @@
+
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, "AT+USOCL=1", &mdata.sem_response,
+			     K_SECONDS(4));
+
+	return ret;
+}
+
+/* reset module */
+void n310_modem_reset(void)
+{
+	int ret = 0, counter = 0;
+
+	pin_init();
+
+	at.CCID = "AT+CCID";
+	at.USORF = "AT+USORF=1";
+	at.NSOSTF = "AT+NSOSTF=1,\"192.168.0.20\",4242,2,4,\"0102AABB\"";
+
+	LOG_INF("Starting modem...");
+
+	static const struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
+		/* stop functionality */
+		SETUP_CMD_NOHANDLE("AT+CFUN=0"),
+		/* extended error numbers */
+		SETUP_CMD_NOHANDLE("AT+CMEE=1"),
+		/* UNC messages for registration */
+		SETUP_CMD_NOHANDLE("AT+CREG=1"),
+		/* enable PDP context */
+		SETUP_CMD_NOHANDLE("AT+CIPCA=1"),
+		/* enable PSM */
+		SETUP_CMD_NOHANDLE("AT+CPSMS=1,,,\"01011111\",\"00000000\""),
+		/* enable HEX mode for +USOWR, +USOST, +USORD and +USORF AT commands. */
+		SETUP_CMD_NOHANDLE("AT+UDCONF=1,1"),
+
+		/* get and store modem info */
+		SETUP_CMD("AT+CGMI", "", on_cmd_atcmdinfo_manufacturer, 0U, ""),
+		SETUP_CMD("AT+CGMM", "", on_cmd_atcmdinfo_model, 0U, ""),
+		SETUP_CMD("AT+CGMR", "", on_cmd_atcmdinfo_revision, 0U, ""),
+		SETUP_CMD("AT+CGSN", "", on_cmd_atcmdinfo_imei, 0U, ""),
+
+		/* enable functionality */
+		SETUP_CMD_NOHANDLE("AT+CFUN=1"),
+	};
+
+	/* reset module */
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, "AT+CFUN=16", &mdata.sem_response,
+			     K_SECONDS(5));
+
+	if (ret < 0) {
+		LOG_ERR("module reset failed: %d", ret);
+		return;
+	}
+
+	modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler, NULL,
+		       0, "AT+CCID", &mdata.sem_response, K_SECONDS(1));
+
+	/* give modem time to start responding after restart */
+	ret = -1;
+
+	while (counter++ < 50 && ret < 0) {
+		k_sleep(K_SECONDS(2));
+		ret = modem_cmd_send(&mdata.context.iface,
+				     &mdata.context.cmd_handler, NULL, 0, "AT",
+				     &mdata.sem_response, K_SECONDS(5));
+		if (ret < 0 && ret != -ETIMEDOUT) {
+			break;
+		}
+	}
+
+	if (ret < 0) {
+		LOG_ERR("Modem wait loop error: %d", ret);
+		return;
+	}
+
+	/* send setup commands */
+	ret = modem_cmd_handler_setup_cmds(&mdata.context.iface,
+					   &mdata.context.cmd_handler,
+					   setup_cmds, ARRAY_SIZE(setup_cmds),
+					   &mdata.sem_response, K_SECONDS(5));
+
+	if (ret < 0) {
+		LOG_ERR("Modem setup cmds error: %d", ret);
+		return;
+	}
+
+	/* register operator automatically */
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, "AT+COPS=0", &mdata.sem_response,
+			     K_SECONDS(120));
+
+	if (ret < 0) {
+		LOG_ERR("AT+COPS error: %d", ret);
+		return;
+	}
+
+	close_socket();
+
+	/* re-init underlying UART comms */
+	ret = modem_iface_uart_init_dev(&mdata.context.iface,
+					CONFIG_MODEM_UBLOX_SARA_N310_UART_NAME);
+
+	if (ret < 0) {
+		LOG_ERR("uart re-init failed");
+		return;
+	}
+
+	LOG_INF("Modem is ready.");
+}
```

Comment:
is it possible to make this function return an int so you can return an error code?